truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1449,6 +1459,315 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
         }
     }
 
+    @CoreMethod(names = "scrub_internal", required = 1, needsBlock = true)
+    public abstract static class ScrubNode extends YieldingCoreMethodNode {
+
+        private static final byte[] SCRUB_REPL_UTF8 = new byte[]{(byte)0xEF, (byte)0xBF, (byte)0xBD};
+        private static final byte[] SCRUB_REPL_ASCII = new byte[]{(byte)'?'};
+        private static final byte[] SCRUB_REPL_UTF16BE = new byte[]{(byte)0xFF, (byte)0xFD};
+        private static final byte[] SCRUB_REPL_UTF16LE = new byte[]{(byte)0xFD, (byte)0xFF};
+        private static final byte[] SCRUB_REPL_UTF32BE = new byte[]{(byte)0x00, (byte)0x00, (byte)0xFF, (byte)0xFD};
+        private static final byte[] SCRUB_REPL_UTF32LE = new byte[]{(byte)0xFD, (byte)0xFF, (byte)0x00, (byte)0x00};
+
+
+        @Child private IsTaintedNode isTaintedNode;
+        @Child private TaintNode taintNode;
+        @Child private CallDispatchHeadNode strCompatAndValidNode;
+        @Child private RopeNodes.MakeConcatNode makeConcatNode;
+
+        public ScrubNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            isTaintedNode = IsTaintedNode.create();
+            taintNode = TaintNode.create();
+            strCompatAndValidNode = DispatchHeadNodeFactory.createMethodCall(context);
+            makeConcatNode = RopeNodesFactory.MakeConcatNodeGen.create(null, null, null);
+        }
+
+        @Specialization
+        public DynamicObject scrubNoBlock(VirtualFrame frame, DynamicObject string, Object repl, NotProvided block) {
+            return scrubDefault(frame, string, repl, nil());
+        }
+
+        @Specialization
+        public DynamicObject scrubDefault(VirtualFrame frame, DynamicObject string, Object repl, DynamicObject block) {
```

Comment:
This is a huge and complex method to compile. Unfortunately it needs the frame to `yield()`.
Maybe some of it could be extracted in smaller TruffleBoundary functions, or the high-level logic be extracted to Ruby.